File: Strand/Data/MedicationStore.swift
import Foundation
import Security

/// One active medication the user chose to record. The name and optional dose/timing note remain inside
/// the device-only secure store; analytics receives only whether any entry has a recent change date.
struct MedicationEntry: Codable, Equatable, Identifiable, Sendable {
    let id: UUID
    var name: String
    var details: String
    /// Optional local ISO day for a medication start or dose change. Editing the text does not
    /// automatically change this clinical-context date.
    var changeDay: String?

    init(id: UUID = UUID(), name: String, details: String = "", changeDay: String? = nil) {
        self.id = id
        self.name = name
        self.details = details
        self.changeDay = changeDay
    }
}

/// Pure policy kept separate from Keychain I/O so the recency boundary is deterministic and testable.
enum MedicationContextPolicy {
    static let recentChangeDays = 14

    static func hasRecentChange(_ entries: [MedicationEntry], asOfDay: String) -> Bool {
        guard let asOf = parseDay(asOfDay) else { return false }
        return entries.contains { entry in
            guard let day = entry.changeDay,
                  let changed = parseDay(day) else { return false }
            let age = calendar.dateComponents([.day], from: changed, to: asOf).day ?? Int.max
            return age >= 0 && age <= recentChangeDays
        }
    }

    static func validDay(_ value: String) -> Bool {
        parseDay(value) != nil
    }

    static func localDayKey(_ date: Date, timeZone: TimeZone = .autoupdatingCurrent) -> String {
        var localCalendar = Calendar(identifier: .gregorian)
        localCalendar.timeZone = timeZone
        let components = localCalendar.dateComponents([.year, .month, .day], from: date)
        guard let year = components.year,
              let month = components.month,
              let day = components.day else {
            return ""
        }
        return String(
            format: "%04d-%02d-%02d",
            locale: Locale(identifier: "en_US_POSIX"),
            year,
            month,
            day
        )
    }

    private static var calendar: Calendar {
        var value = Calendar(identifier: .gregorian)
        value.timeZone = TimeZone(secondsFromGMT: 0)!
        return value
    }

    private static func parseDay(_ value: String) -> Date? {
        guard value.count == 10 else { return nil }
        let parts = value.split(separator: "-", omittingEmptySubsequences: false)
        guard parts.count == 3,
              let year = Int(parts[0]),
              let month = Int(parts[1]),
              let day = Int(parts[2]) else { return nil }
        let components = DateComponents(
            calendar: calendar,
            timeZone: calendar.timeZone,
            year: year,
            month: month,
            day: day
        )
        guard let date = calendar.date(from: components) else { return nil }
        let roundTrip = calendar.dateComponents([.year, .month, .day], from: date)
        return roundTrip.year == year && roundTrip.month == month && roundTrip.day == day
            ? date
            : nil
    }
}

/// Device-only Keychain storage for active medications. This data is intentionally outside the app
/// database, UserDefaults, portable backups, self-hosted sync, Coach context, analytics, and logs.
enum MedicationStore {
    private static let service = "com.noop.medication-context"
    private static let account = "active-medications"
    private static let maximumEntries = 30
    private static let maximumNameCharacters = 120
    private static let maximumDetailsCharacters = 240

    private static var query: [String: Any] {
        [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service,
            kSecAttrAccount as String: account,
        ]
    }

    static var active: [MedicationEntry] {
        var lookup = query
        lookup[kSecReturnData as String] = kCFBooleanTrue
        lookup[kSecMatchLimit as String] = kSecMatchLimitOne
        var item: CFTypeRef?
        let status = SecItemCopyMatching(lookup as CFDictionary, &item)
        guard status == errSecSuccess,
              let data = item as? Data,
              let decoded = try? JSONDecoder().decode([MedicationEntry].self, from: data) else {
            return []
        }
        return normalized(decoded)
    }

    static var hasRecentChange: Bool {
        MedicationContextPolicy.hasRecentChange(
            active,
            asOfDay: MedicationContextPolicy.localDayKey(Date())
        )
    }

    @discardableResult
    static func upsert(_ entry: MedicationEntry) -> Bool {
        guard let clean = normalized(entry) else { return false }
        var entries = active
        if let index = entries.firstIndex(where: { $0.id == clean.id }) {
            entries[index] = clean
        } else {
            guard entries.count < maximumEntries else { return false }
            entries.append(clean)
        }
        return save(entries)
    }

    @discardableResult
    static func remove(id: UUID) -> Bool {
        let entries = active
        let updated = entries.filter { $0.id != id }
        guard updated.count != entries.count else { return true }
        return save(updated)
    }

    private static func save(_ entries: [MedicationEntry]) -> Bool {
        guard let data = try? JSONEncoder().encode(normalized(entries)) else { return false }
        let status = SecItemUpdate(
            query as CFDictionary,
            [kSecValueData as String: data] as CFDictionary
        )
        if status == errSecSuccess { return true }
        guard status == errSecItemNotFound else { return false }

        var attributes = query
        attributes[kSecValueData as String] = data
        attributes[kSecAttrAccessible as String] = kSecAttrAccessibleAfterFirstUnlockThisDeviceOnly
        return SecItemAdd(attributes as CFDictionary, nil) == errSecSuccess
    }

    private static func normalized(_ entries: [MedicationEntry]) -> [MedicationEntry] {
        var seen = Set<UUID>()
        return entries.compactMap(normalized)
            .filter { seen.insert($0.id).inserted }
            .prefix(maximumEntries)
            .map { $0 }
    }

    private static func normalized(_ entry: MedicationEntry) -> MedicationEntry? {
        let name = String(
            entry.name.trimmingCharacters(in: .whitespacesAndNewlines)
                .prefix(maximumNameCharacters)
        )
        guard !name.isEmpty else { return nil }
        let details = String(
            entry.details.trimmingCharacters(in: .whitespacesAndNewlines)
                .prefix(maximumDetailsCharacters)
        )
        let changeDay = entry.changeDay.flatMap {
            MedicationContextPolicy.validDay($0) ? $0 : nil
        }
        return MedicationEntry(id: entry.id, name: name, details: details, changeDay: changeDay)
    }
}

File: android/app/src/main/java/com/noop/data/MedicationStore.kt
package com.noop.data

import android.content.Context
import org.json.JSONArray
import org.json.JSONObject
import java.time.LocalDate
import java.time.temporal.ChronoUnit
import java.util.UUID

/** One current medication. Text stays encrypted locally; analytics receives no name, dose, or timing. */
data class MedicationEntry(
    val id: String = UUID.randomUUID().toString(),
    val name: String,
    val details: String = "",
    /** Optional ISO local day on which the medication was started or its dose changed. */
    val changeDay: String? = null,
)

/** Pure, testable recency rule for medication context. */
object MedicationContextPolicy {
    const val RECENT_CHANGE_DAYS = 14L

    fun hasRecentChange(
        entries: List<MedicationEntry>,
        asOfDay: LocalDate,
    ): Boolean = entries.any { entry ->
        val changed = entry.changeDay?.let { parseDay(it) } ?: return@any false
        val age = ChronoUnit.DAYS.between(changed, asOfDay)
        age in 0L..RECENT_CHANGE_DAYS
    }

    fun validDay(value: String): Boolean = parseDay(value) != null

    private fun parseDay(value: String): LocalDate? =
        runCatching { LocalDate.parse(value) }.getOrNull()
}

/**
 * Keystore-backed active-medication storage. It is deliberately outside Room, ordinary preferences,
 * portable backup, self-hosted sync, Coach context, analytics, and logs.
 */
object MedicationStore {
    private const val FILE_NAME = "noop_medication_secure_prefs"
    private const val KEY_ENTRIES = "active_medications"
    private const val MAXIMUM_ENTRIES = 30
    private const val MAXIMUM_NAME_CHARACTERS = 120
    private const val MAXIMUM_DETAILS_CHARACTERS = 240
    private val lock = Any()
    @Volatile private var cachedEntries: List<MedicationEntry>? = null

    fun active(context: Context): List<MedicationEntry> {
        cachedEntries?.let { return it }
        return synchronized(lock) {
            cachedEntries ?: runCatching {
                val raw = SecurePrefs.of(context, FILE_NAME).getString(KEY_ENTRIES, null)
                    ?: return@runCatching emptyList()
                decode(raw)
            }.getOrDefault(emptyList()).also { cachedEntries = it }
        }
    }

    fun hasRecentChange(
        context: Context,
        asOfDay: LocalDate = LocalDate.now(),
    ): Boolean = MedicationContextPolicy.hasRecentChange(active(context), asOfDay)

    fun upsert(context: Context, entry: MedicationEntry): Boolean = synchronized(lock) {
        val clean = normalize(entry) ?: return false
        val entries = active(context).toMutableList()
        val index = entries.indexOfFirst { it.id == clean.id }
        if (index >= 0) {
            entries[index] = clean
        } else {
            if (entries.size >= MAXIMUM_ENTRIES) return false
            entries.add(clean)
        }
        return save(context, entries)
    }

    fun remove(context: Context, id: String): Boolean = synchronized(lock) {
        val entries = active(context)
        val updated = entries.filterNot { it.id == id }
        if (updated.size == entries.size) return true
        return save(context, updated)
    }

    private fun save(context: Context, entries: List<MedicationEntry>): Boolean {
        val clean = normalize(entries)
        val saved = runCatching {
            SecurePrefs.of(context, FILE_NAME)
            .edit()
            .putString(KEY_ENTRIES, encode(clean))
            .commit()
        }.getOrDefault(false)
        if (saved) cachedEntries = clean
        return saved
    }

    private fun normalize(entries: List<MedicationEntry>): List<MedicationEntry> {
        val seen = HashSet<String>()
        return entries.asSequence()
            .mapNotNull(::normalize)
            .filter { seen.add(it.id) }
            .take(MAXIMUM_ENTRIES)
            .toList()
    }

    private fun normalize(entry: MedicationEntry): MedicationEntry? {
        val id = runCatching { UUID.fromString(entry.id).toString() }.getOrNull() ?: return null
        val name = entry.name.trim().take(MAXIMUM_NAME_CHARACTERS)
        if (name.isEmpty()) return null
        val details = entry.details.trim().take(MAXIMUM_DETAILS_CHARACTERS)
        val changeDay = entry.changeDay?.takeIf(MedicationContextPolicy::validDay)
        return MedicationEntry(id, name, details, changeDay)
    }

    private fun encode(entries: List<MedicationEntry>): String {
        val array = JSONArray()
        entries.forEach { entry ->
            array.put(JSONObject().apply {
                put("id", entry.id)
                put("name", entry.name)
                put("details", entry.details)
                put("changeDay", entry.changeDay ?: JSONObject.NULL)
            })
        }
        return array.toString()
    }

    private fun decode(raw: String): List<MedicationEntry> {
        val array = JSONArray(raw)
        val entries = ArrayList<MedicationEntry>(array.length())
        for (index in 0 until array.length()) {
            val value = array.optJSONObject(index) ?: continue
            entries.add(
                MedicationEntry(
                    id = value.optString("id"),
                    name = value.optString("name"),
                    details = value.optString("details"),
                    changeDay = if (value.isNull("changeDay")) null else value.optString("changeDay"),
                )
            )
        }
        return normalize(entries)
    }
}
